### PR TITLE
[Rust] A fiber based async mechanism design.

### PR DESCRIPTION
**Pull request by MediosZ**
Signed-off-by: Tricster <mediosrity@gmail.com>

This pr aims to add async capability for rust sdk, so that the sdk works in an async way and users can define async host functions.

### CODE REVIEW

File: bindings/rust/wasmedge-sys/Cargo.toml [hunk 1]
```diff
@@ -21,11 +21,14 @@ rand = "0.8.4"
 thiserror = "1.0.30"
 wasmedge-macro = {path = "../wasmedge-macro", version = "0.1"}
 wasmedge-types = {path = "../wasmedge-types", version = "0.3"}
-
+wasmtime-fiber = "1"
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/Cargo.toml:24]: Looks like we should have an empty line before `[build-dependencies]`.

File: bindings/rust/wasmedge-sys/examples/async_host_func2.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+use wasmedge_sys::{Executor, FuncType, Function, Store};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // create a FuncType
+    let func_ty = FuncType::create(vec![], vec![])?;
+    let mut store = Store::create().expect("123");
+    dbg!(&store.async_state.current_poll_cx.get());
+    // println!("{:?}", store.async_cx().unwrap());
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/examples/async_host_func2.rs:9]: Redundant line.

File: bindings/rust/wasmedge-sys/examples/async_host_func2.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                // tx.send(()).unwrap();
+
+                // Wrap the future with a `Timeout` set to expire in 10 milliseconds.
+                // if let Err(_) = tokio::time::timeout(std::time::Duration::from_millis(100), async {
+                //     tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                // })
+                // .await
+                // {
+                //     println!("did not receive value within 100 ms");
+                // }
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/examples/async_host_func2.rs:27]: Redundant line.

File: bindings/rust/wasmedge-sys/examples/async_host_func2.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                Ok(vec![])
+            })
+        }),
+        Some(&mut store),
+        0,
+    )?;
+
+    // run this function
+    let mut executor = Executor::create(None, None)?;
+    // println!("{:?}", store.async_cx().unwrap());
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/examples/async_host_func2.rs:39]: Redundant line.

File: bindings/rust/wasmedge-sys/examples/async_host_func3.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    let config = Config::create()?;
+    let mut store = Store::create().expect("Unable to create store");
+    let mut vm = Vm::create(Some(config), Some(&mut store))?;
+
+    // dbg!(&store);
+    // dbg!(&store.async_state.current_poll_cx.get());
+    // dbg!(&store.async_state.current_suspend.get());
+    // unsafe {
+    //     dbg!(*(store.async_state.current_poll_cx.get()));
+    // }
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/examples/async_host_func3.rs:78]: Redundant line.

File: bindings/rust/wasmedge-sys/examples/async_run_func.rs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    // instantiate
+    let result = vm.instantiate();
+    assert!(result.is_ok());
+
+    // async run function
+    let fut1 = vm.run_function_async2(String::from("fib"), vec![WasmValue::from_i32(20)]);
+
+    let fut2 = vm.run_function_async2(String::from("fib"), vec![WasmValue::from_i32(5)]);
+    // println!("{:?}", fut1.await);
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/examples/async_run_func.rs:52]: Redundant line.

File: bindings/rust/wasmedge-sys/src/async_env.rs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        unsafe {
+            *self.0 = self.1;
+        }
+    }
+}
+
+pub(crate) struct FiberFuture<'a> {
+    fiber: Fiber<'a, Result<(), ()>, (), Result<(), ()>>,
+    current_poll_cx: *mut *mut Context<'static>,
+    // engine: Engine,
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/async_env.rs:22]: Why does this comment out?

**Review by hydai — CHANGES_REQUESTED** [review]

File: bindings/rust/wasmedge-sys/examples/async_host_func2.rs (hunk 2 above)
Comment (MediosZ) [reply bindings/rust/wasmedge-sys/examples/async_host_func2.rs:9]: This is still a draft, and all these will be removed later.

Comment (hydai) [reply bindings/rust/wasmedge-sys/examples/async_host_func2.rs:9]: Ah, sorry, not noticed this is still a draft.

File: bindings/rust/wasmedge-sys/Cargo.toml (hunk 1 above)
Comment (MediosZ) [reply bindings/rust/wasmedge-sys/Cargo.toml:24]: Sure

File: bindings/rust/wasmedge-sys/src/async_env.rs (hunk 7 above)
Comment (MediosZ) [reply bindings/rust/wasmedge-sys/src/async_env.rs:22]: Cause this was added in previous versions and not needed anymore, will delete it.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 8]
```diff
@@ -360,6 +363,25 @@ impl Vm {
         self.run_function_async(func_name, params)
     }
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:364]: Please replace the original `run_wasm_from_file_async` with `run_wasm_from_file_async2`, but keep using the original API name `run_wasm_from_file_async`. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 9]
```diff
@@ -430,6 +452,25 @@ impl Vm {
...
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:453]: Please replace the origin `run_wasm_from_bytes_async` with the new `run_wasm_from_bytes_async2`, but keep using the original API name `run_wasm_from_bytes_async`. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 10]
```diff
@@ -532,7 +592,7 @@ impl Vm {
     /// # Error
     ///
     /// If fail to load, then an error is returned.  The loaded module is not validated.
-    pub fn load_wasm_from_bytes(&mut self, bytes: &[u8]) -> WasmEdgeResult<()> {
+    pub fn load_wasm_from_bytes(&self, bytes: &[u8]) -> WasmEdgeResult<()> {
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:595]: Please still use '&mut self` as the design of the current `sys` APIs tried to "express" the mutability of the relevant C-API.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 11]
```diff
@@ -511,7 +571,7 @@ impl Vm {
     /// # Error
     ///
     /// If fail to load, then an error is returned.
-    pub fn load_wasm_from_module(&mut self, module: &Module) -> WasmEdgeResult<()> {
+    pub fn load_wasm_from_module(&self, module: &Module) -> WasmEdgeResult<()> {
         unsafe {
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:575]: Please still use '&mut selfas the design of the currentsys` APIs tried to "express" the mutability of the relevant C-API.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 12]
```diff
@@ -360,6 +363,25 @@ impl Vm {
...
+        // validate
+        self.validate()?;
+
+        // instantiate
+        self.instantiate()?;
+
+        // invoke
+        self.run_function_async2(func_name, params).await
+    }
+
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:384]: Please replace the original `run_wasm_from_file_async` with `run_wasm_from_file_async2`, but keep using the original API name `run_wasm_from_file_async`. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 13]
```diff
@@ -430,6 +452,25 @@ impl Vm {
...
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:473]: Please replace the origin `run_wasm_from_bytes_async` with the new `run_wasm_from_bytes_async2`, but keep using the original API name `run_wasm_from_bytes_async`. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 14]
```diff
@@ -511,7 +571,7 @@ impl Vm {
     /// # Error
     ///
     /// If fail to load, then an error is returned.
-    pub fn load_wasm_from_module(&mut self, module: &Module) -> WasmEdgeResult<()> {
+    pub fn load_wasm_from_module(&self, module: &Module) -> WasmEdgeResult<()> {
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:574]: Please still use '&mut selfas the design of the currentsys` APIs tried to "express" the mutability of the relevant C-API.

File: bindings/rust/wasmedge-sys/src/vm.rs (hunk 10 above)
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:595]: Same issue as above.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 15]
```diff
@@ -554,7 +614,7 @@ impl Vm {
     /// # Error
     ///
     /// If fail to load, then an error is returned.  The loaded module is not validated.
-    pub fn load_wasm_from_file(&mut self, path: impl AsRef<Path>) -> WasmEdgeResult<()> {
+    pub fn load_wasm_from_file(&self, path: impl AsRef<Path>) -> WasmEdgeResult<()> {
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:617]: Same issue as above.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 16]
```diff
@@ -572,7 +632,7 @@ impl Vm {
     /// # Error
     ///
     /// If fail to validate, then an error is returned.
-    pub fn validate(&mut self) -> WasmEdgeResult<()> {
+    pub fn validate(&self) -> WasmEdgeResult<()> {
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:635]: Same issue as above.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 17]
```diff
@@ -586,7 +646,7 @@ impl Vm {
     /// # Error
     ///
     /// If fail to instantiate, then an error is returned.
-    pub fn instantiate(&mut self) -> WasmEdgeResult<()> {
+    pub fn instantiate(&self) -> WasmEdgeResult<()> {
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:649]: Same issue as above.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 18]
```diff
@@ -500,6 +541,25 @@ impl Vm {
...
+        // validate
+        self.validate()?;
+
+        // instantiate
+        self.instantiate()?;
+
+        // invoke
+        self.run_function_async2(func_name, params).await
+    }
+
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:562]: Please replace the origin `run_wasm_from_module_async` with the new `run_wasm_from_module_async2`, but keep using the original API name `run_wasm_from_module_async`. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 19]
```diff
@@ -673,6 +733,15 @@ impl Vm {
...
     }
+    pub async fn run_function_async2(
+        &self,
+        func_name: impl AsRef<str> + Send,
+        params: impl IntoIterator<Item = WasmValue> + Send,
+    ) -> WasmEdgeResult<Vec<WasmValue>> {
+        FiberFuture::on_fiber(|| self.run_function(func_name, params))
+            .await
+            .unwrap()
+    }
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:744]: Please replace the original `run_function_async` with the new `run_function_async2`, but keep using the original API name `run_function_async'. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 20]
```diff
@@ -763,7 +832,16 @@ impl Vm {
...
+    pub async fn run_registered_function_async2(
+        &self,
+        mod_name: impl AsRef<str> + Send,
+        func_name: impl AsRef<str> + Send,
+        params: impl IntoIterator<Item = WasmValue> + Send,
+    ) -> WasmEdgeResult<Vec<WasmValue>> {
+        FiberFuture::on_fiber(|| self.run_registered_function(mod_name, func_name, params))
+            .await
+            .unwrap()
+    }
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/vm.rs:844]: Please replace the original `run_registered_function_async` with the new `run_registered_function_async2`, but keep using the original API name `run_registered_function_async`. Thanks!

File: bindings/rust/wasmedge-sys/src/lib.rs [hunk 21]
```diff
@@ -238,6 +239,18 @@ pub type BoxedFn = Box<
...
+            &CallingFrame,
+            Vec<WasmValue>,
+            *mut std::os::raw::c_void,
+        ) -> Box<
+            dyn std::future::Future<Output = Result<Vec<WasmValue>, error::HostFuncError>> + Send,
+        > + Send
+        + Sync
+        + 'static,
+>;
+
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/lib.rs:253]: Please move the type alias declaration to the location just before line 265, as the type alias is paired with `ASYNC_STATE`.

File: bindings/rust/wasmedge-sys/src/instance/function.rs [hunk 22]
```diff
@@ -321,6 +321,30 @@ impl Function {
         }
     }
 
+    pub fn create_async<T>(
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/instance/function.rs:324]: Please add doc comments for this new API. As for the format of the API doc comments, please refer to the existed. Thanks!

File: bindings/rust/wasmedge-sys/src/instance/function.rs [hunk 23]
```diff
@@ -415,6 +439,16 @@ impl Function {
         engine.run_func(self, args)
     }
 
+    pub async fn call_async<E: Engine + Send + Sync>(
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/instance/function.rs:442]: Please add doc comments for this new API. As for the format of the API doc comments, please refer to the existed. Thanks!

File: bindings/rust/wasmedge-sys/src/async_env.rs [hunk 24]
```diff
@@ -0,0 +1,138 @@
+use std::{
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/src/async_env.rs:1]: Please add a summary for this new module. As for the format, please refer to the existed. Thanks!

Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/async_env.rs:1]: In addition, please also add doc comments for all `public` types and interfaces. Thanks!

File: bindings/rust/wasmedge-sys/examples/async_run_wasm_from_file.rs [hunk 25]
```diff
@@ -0,0 +1,33 @@
+use wasmedge_sys::{Config, Vm, WasmValue};
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/examples/async_run_wasm_from_file.rs:1]: Please add the necessary description of how to build/run this example. If there was any assumption, please also make it clear to users. Thanks!

File: bindings/rust/wasmedge-sys/examples/async_run_wasm_from_bytes.rs [hunk 26]
```diff
@@ -0,0 +1,70 @@
+use wasmedge_sys::{Config, Store, Vm, WasmValue};
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/examples/async_run_wasm_from_bytes.rs:1]: Please add the necessary description of how to build/run this example. If there was any assumption, please also make it clear to users. Thanks!

File: bindings/rust/wasmedge-sys/examples/async_run_named_func.rs [hunk 27]
```diff
@@ -0,0 +1,45 @@
+use wasmedge_sys::{Config, Store, Vm, WasmValue};
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/examples/async_run_named_func.rs:1]: Please add the necessary description of how to build/run this example. If there was any assumption, please also make it clear to users. Thanks!

File: bindings/rust/wasmedge-sys/examples/async_run_func.rs [hunk 28]
```diff
@@ -0,0 +1,62 @@
+use wasmedge_sys::{Config, Loader, Store, Vm, WasmValue};
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/examples/async_run_func.rs:1]: Please add the necessary description of how to build/run this example. If there was any assumption, please also make it clear to users. Thanks!

File: bindings/rust/wasmedge-sys/examples/async_host_func2.rs [hunk 29]
```diff
@@ -0,0 +1,39 @@
+use wasmedge_sys::{Executor, FuncType, Function, Store};
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/examples/async_host_func2.rs:1]: Please add the necessary description of how to build/run this example. If there was any assumption, please also make it clear to users. Thanks!

File: bindings/rust/wasmedge-sys/Cargo.toml [hunk 30]
```diff
@@ -21,11 +21,15 @@ rand = "0.8.4"
 thiserror = "1.0.30"
 wasmedge-macro = {path = "../wasmedge-macro", version = "0.1"}
 wasmedge-types = {path = "../wasmedge-types", version = "0.3"}
+wasmtime-fiber = "1"
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/Cargo.toml:24]: What about using `wasmtime-fiber = "2"`? Seems the latest version of `wasmtime-fiber` is `2.0+`. Thanks!

**Review by apepkuss — CHANGES_REQUESTED** [review]

**Comment by apepkuss** [comment]
Hi @MediosZ,
According to the CI results, you need to follow the steps below to check and fix the clipply issues in this PR:

- Use `rustup update nightly` command to update `Clippy` to the latest version first; 
- Then, run `cargo +nightly clippy --profile test --lib --examples -- -D warnings -D clippy::dbg_macro` command in the root directory of `bindings/rust` in your local environment;
-  Fix the clippy issues detected.

Thanks!

**Comment by apepkuss** [comment]
@MediosZ According to the CI results, some clippy issues are still not fixed. Please fix them. BTW, the CI workflow uses `clippy 0.1.66 (9565dfe 2022-10-28)`, please guarantee that you have the same version of `Cargo Clippy` in your local system. Thanks!

**Comment by MediosZ** [comment]
The error is following:
```
error: variable does not need to be mutable
   --> wasmedge-sdk/src/vm.rs:130:9
    |
130 |         mut self,
    |         ----^^^^
    |         |
    |         help: remove this `mut`
    |
    = note: `-D unused-mut` implied by `-D warnings`
```
I left wasmedge-sdk untouched in this pr, but if this error needs to be fixed in this pr, I can do it
@apepkuss

**Comment by apepkuss** [comment]
> The error is following:
> 
> ```
> error: variable does not need to be mutable
>    --> wasmedge-sdk/src/vm.rs:130:9
>     |
> 130 |         mut self,
>     |         ----^^^^
>     |         |
>     |         help: remove this `mut`
>     |
>     = note: `-D unused-mut` implied by `-D warnings`
> ```
> 
> I left wasmedge-sdk untouched in this pr, but if this error needs to be fixed in this pr, I can do it @apepkuss

@MediosZ Thanks for the explanation. Please try to fix it first. If the fix would cause other failures, such as failed test cases, please leave it there. Thanks a lot!

**Review by apepkuss — APPROVED** [review]

**Review by apepkuss — APPROVED** [review]

File: bindings/rust/wasmedge-sys/examples/async_run_func.rs [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+
+    // instantiate
+    let result = vm.instantiate();
+    assert!(result.is_ok());
+
+    // async run function
+    let fut1 = vm.run_function_async2(String::from("fib"), vec![WasmValue::from_i32(20)]);
+
+    let fut2 = vm.run_function_async2(String::from("fib"), vec![WasmValue::from_i32(5)]);
+    // println!("{:?}", fut1.await);
```
Comment (apepkuss) [inline bindings/rust/wasmedge-sys/examples/async_run_func.rs:62]: Please remove the commented code if it is not used. Thanks!

**Review by apepkuss — APPROVED** [review]
Please fix the only issue. Thanks!

**Review by apepkuss — APPROVED** [review]

**Review by hydai — APPROVED** [review]

**Pull request merged**